lessons/intro/regex/index.md

```diff
@@ -0,0 +1,268 @@
+která je možné v případě potřeby zapnout.
+
+* `re.DEBUG` zobrazí ladící hlášky týkající se zpracování regulárního výrazu
+* `re.IGNORECASE` (zkráceně `re.I`) způsobí, že regulární výraz nebude brát ohled na velikost písmen
+* `re.MULTILINE` (zkráceně `re.M`) umožní zpracování víceřádkových řetězců. `^` a `$` začnou označovat začátek a konec každého řádku místo začátku a konce celého řetězce.
+* `re.DOTALL` (zkráceně `re.S`) způsobí, že se tečka stane zástupným znakem i za nový řádek `\n`, což v základním nastavení není.
+* … a další k nalezení v dokumentaci.
+
+Nejlepší je předávat flagy jako pojmenovaný argument výše zmíněným funkcím. Vyhneme se tím chybám vzniklým kvůli různým počtům pozičních argumentů funkcí v modulu `re`.
+Pokud jich potřebujeme předat více najednou, nakombinujeme je pomocí operátoru `|`.
+
+```pycon
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři HODINY.")
+['ny', 'ty']
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři HODINY.", flags=re.IGNORECASE)
+['ny', 'ty', 'NY']
+>>> re.findall(r".y", "Naše staré hodiny bijí čtyři HODINY.", flags=re.IGNORECASE|re.DEBUG)
+ANY None
+LITERAL 121
+
+0. INFO 4 0b0 2 2 (to 5)
+5: ANY
+6. LITERAL_UNI_IGNORE 0x79 ('y')
+8. SUCCESS
+['ny', 'ty', 'NY']
+```
+
+## Optimalizace
+
+Jednou z možností optimalizace fungování regulárních výrazů je jejich kompilace. Nejdříve si regulární výraz zkompilujete a pak místo
+funkcí z modulu `re` použijete stejnojmenné metody tohoto nového objektu.
+
+```pycon
+>>> regex = re.compile(r".y", flags=re.IGNORECASE)
+>>> regex.findall("Naše staré hodiny bijí čtyři HODINY.")
+['ny', 'ty', 'NY']
+```
+
+Kompilace se ale děje i automaticky na pozadí při používání funkcí, takže pokud používáte jen několik stále se opakujících
+regulárních výrazů, není třeba si je předem ručně kompilovat.
+
+## Příklad z úvodu lekce
+
+Teď se můžeme společně podívat, jestli se nám podaří rozšifrovat kryptický zápis z úvodu této lekce. V ideálním případě bychom k tomu měli znát vše potřebné.
+
+```
+r"^\w+([\.-]?\w)+@\w+([\.]?\w)+(\.[a-zA-Z]{2,3})+$"
+```
+
+Vezmeme to postupně:
+* `^` označení začátku hledaného řetězce
+* `\w+` 1 nebo více alfanumerických znaků
+* `([\.-]?\w)+` skupina alfanumerických znaků, která může začínat tečkou nebo pomlčkou a může se jednou nebo vícekrát opakovat
+* @ znak zavináče — nemá žádný speciální význam
+* `\w+` 1 nebo více alfanumerických znaků
+* `([\.]?\w)+` skupina alfanumerických znaků, která může začínat tečkou a může se jednou nebo vícekrát opakovat
+* `(\.[a-zA-Z]{2,3})+` skupina obsahující dvě nebo tři malá/velká písmena anglické abecedy nebo číslice, která musí začínat tečkou a může se jednou nebo vícekrát opakovat
+* `$` tady by měl validovaný řetězec skončit a neměl by obsahovat nic navíc
+
+Pokud to ještě nedává smysl, jedná se o regulární výraz pro validaci emailových adres. Není ovšem bez chyby, protože nám nedovolí použít před zavináčem znak `+`, který je možnou validní součástí emailové adresy.
```

Comment:
Považuji tenhle příklad za trochu nešťastný. Tedy máte výraz, který validuje něco komplexního, ale hehe, je vlastně špatně. A taky je špatně z jiných zde neuvedených důvodů. https://emailregex.com/

Reply:
Příklad bych nechal a doplnil do popisu, že udělat přesný regex pro email vlastně ani nejde a odkazem na emailregex.com bych ukázal, jak komplexní to ve výsledku může být.

Reply:
Ukázal bych (taky) nějaký příklad, jak vytáhnout třeba username z tohohle emailu.